Bluetooth: host: Fix potential deadlocks with the TX thread and system workqueue

jhedberg opened the pull request:
Three related patches, the essence is in the second one and described fairly well in its commit message:

This is a moderate redesign of the pending TX packet handling that aims to eliminate potential deadlocks between the TX thread and the system workqueue thread. The main changes are:
    
 - TX context (`bt_conn_tx`) is allocated during buffer allocation, i.e. not in the TX thread.   
 - We don't allocate a TX context unless there's an associated callback. When there's no callback simple integer counters are used for tracking.
 - The TX thread is no longer responsible for TX callbacks or scheduling of TX callbacks. Instead, the callbacks get directly scheduled (`k_work_submit`) from the RX priority thread.
 - The `CONFIG_BT_CONN_TX_MAX` defaults to `CONFIG_BT_L2CAP_TX_BUF_COUNT` and in most cases won't need changing.

Fixes #20938

Commit:
Bluetooth: host: Shrink size of struct bt_conn_tx

The `node` and `work` members are never used simultaneously.
Additionally k_work already has built-in support for being in a linked
list, however a union makes this change a bit cleaner.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: host: Fix deadlocks with pending TX packet handling

This is a moderate redesign of the pending TX packet handling that
aims to eliminate potential deadlocks between the TX thread and the
system workqueue thread. The main changes are:

 - TX context (bt_conn_tx) is allocated during buffer allocation, i.e.
   not in the TX thread.

 - We don't allocate a TX context unless there's an associated
   callback. When there's no callback simple integer counters are used
   for tracking.

 - The TX thread is no longer responsible for TX callbacks or
   scheduling of TX callbacks. Instead, the callbacks get directly
   scheduled (k_work_submit) from the RX priority thread.

 - CONFIG_BT_CONN_TX_MAX defaults to CONFIG_BT_L2CAP_TX_BUF_COUNT,
   and in most cases wont need changing. The value now only indicates
   how many pending packets with a callback are possible.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: host: Move system workqueue special case to the right place

Now that we've removed the TX allocation dependency from the TX thread
we no longer have the need to do special-casing for the system
workqueue when allocating buffers. Instead, we do have to special-case
the system workqueue when allocating TX contexts since the system
workqueue is the only place where they get freed up.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Vudentz reviewed (approved):
Tested with ipsp ping and had the same result.

Commit:
Bluetooth: L2CAP: Process fixed channels in the RX thread

Now that the TX path and TX context (bt_conn_tx) has been redesigned
to free the contexts always in the system workqueue, it means the
system workqueue is the only context where their allocation may also
fail. This is particularly problematic with us having all L2CAP
channels (fixed & CoC alike) deferred to the system workqueue. It is
especially bad for fixed channels where being able to send responses
for SM, L2CAP signaling and ATT is critical to avoid timeouts for the
connection.

This patch moves the processing of all fixed L2CAP channels back to
the RX thread, thereby making it possible (and safe) to block while
waiting for a TX context to become available.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

joerchan commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -115,10 +116,10 @@ struct bt_conn {
-	/* Acknowledged but not yet notified TX packets */
-	struct k_fifo		tx_notify;
+	/* Sent but not acknowledged TX packets without a callback */
+	u32_t                   pending_no_cb;
```
Could we extend this comment to say that this is before the tx_pending?

i.e /* Sent but not acknowledged TX packets without a callback before the first element  with a callback */ ?

jhedberg replied:
sure, that does indeed make it clearer

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1707,7 +1703,8 @@ void bt_l2cap_recv(struct bt_conn *conn, struct net_buf *buf)
 
-	l2cap_chan_recv_queue(chan, buf);
+	l2cap_chan_recv(chan, buf);
+	net_buf_unref(buf);
```
This change confuses me somewhat. Where was this done previously for the static channels?

jhedberg replied:
In the k_work callback that `l2cap_chan_recv_queue` schedules, i.e. `l2cap_rx_process()`, which then eventually called `l2cap_chan_recv()`

jhedberg commented on include/bluetooth/l2cap.h:
```diff
@@ -132,6 +129,9 @@ struct bt_l2cap_le_chan {
 	struct net_buf			*_sdu;
 	u16_t				_sdu_len;
+
+	struct k_work			rx_work;
```
This is part of the cleanup that I filed an issue for. *Most* of this trust can/should be compiled out without DYNAMIC_CHANNEL, but since I wanted to keep this PR minimal (due to the upcoming release I decided to leave this detail alone for now.
